src/System.Private.ServiceModel/tools/PRService/pr.ashx

```diff
@@ -19,92 +19,135 @@
+        context.Response.Write(string.Format("branch = {0}<br/>", HttpUtility.HtmlEncode(branchString) ?? "unspecified"));
+
+        // Check for prerequisites needed before this script can execute
+        string gitRepoPath = GetRepoPath(repoId);
+        success = CheckPrerequisites(gitRepoPath, result);
+
+        if (!success)
+        {
+            // CheckPrerequisites failed, this is a server issue
+            context.Response.StatusCode = 500;
+            context.Response.Write(HttpUtility.HtmlEncode(result.ToString()));
             return;
         }
 
-        context.Response.Write(string.Format("PR = {0}<br/>", HttpUtility.HtmlEncode(prString) ?? "Unspecified"));
-        context.Response.Write(string.Format("branch = {0}<br/>", HttpUtility.HtmlEncode(branchString) ?? "Unspecified"));
+        // Cleanup all branches except master from the repo - this shouldn't fail, and if it does, there's something wrong with the repo
+        success = CleanupBranches(gitRepoPath, result);
+
+        if (!success)
+        {
+            // CleanupMergedBranches failed, this is a server issue
+            context.Response.StatusCode = 500;
+            context.Response.Write(HttpUtility.HtmlEncode(result.ToString()));
+            return;
+        }
 
         // Do stuff
+        success = false;
 
         uint pr;
         if (!string.IsNullOrEmpty(prString) && uint.TryParse(prString, out pr))
         {
-            success = SyncToPr(pr, result);
+            success = SyncToPr(pr, gitRepoPath, result);
         }
         else if (!string.IsNullOrEmpty(branchString))
         {
-            // We only want to support limited branches as we only use this for PR purposes
-            // This should be opened up later once we decide we need to (e.g., for testing release branches), but for now, 
-            // whitelist what we need
-            // for example, we could use the output from "git branch -r" to figure out what branches are valid
+            // Validate input here, because we don't want someone passing in an arbitrary branch
+            // Obtain a list of valid remote branches and allow sync to one of these only
+
+            string[] branches;
+            success = GetBranches(gitRepoPath, true, out branches, result);
 
-            switch (branchString)
+            if (success)
             {
-                case "master":
-                case "origin/master":
-                case "origin/infrastructure":
-                    success = SyncToBranch(branchString, result);
-                    break;
-                default:
+                success = false;
+                for (int i = 0; i < branches.Length; i++)
+                {
+                    // git branches are case sensitive 
+                    if (string.Equals(branchString, branches[i]))
+                    {
+                        success = SyncToBranch(branchString, gitRepoPath, result);
```

Comment:
Failure here plus the break will lead people to think the branch is unsupported.  It would be better to write 'result' into the response and to use a different error message than the one below.